### ISSUE TITLE
Players losing personal saves

### ISSUE THREAD

**Issue by jaj22** [official-antistasi-community/A3-Antistasi#1031]
*Version:* Unstable, probably since ~2.2 at least

*Mods:* Probably doesn't matter.

*Environment*: MP dedi, maybe other MP

***Issue:***
Players sometimes lose their rank and money on load. Loadouts are still loaded. Not sure about personal garage yet. Being (one of?) the first players to connect may be a risk factor. Not easily replicable.

In this issue, the load player command is correctly sent to the server, and apparently the data is found in the server's profilenamespace. It's not yet clear whether the data has been correctly saved in the first place.

This is distinct from save loss caused by the player changing UID or saving/disconnecting while in the wrong body (latter should be fixed now).

**Comment by jaj22** [comment]
#1032 seems to resolve the problem, so apparently using global setVariables from both client and server is undefined behaviour. Will attempt to write a slightly more coherent fix.

**Comment by Mirv** [comment]
My guy from 2 days ago lost rank/score/garage

**Comment by Mirv** [comment]
@jaj22  ... do we think the issue is file writing, file max size, file loading, file var overwriting, memory over load?

I'm willing to help isolate it & do a work around ...

Senario A:

I create a script asking the player to confirm the information saved is correct.

- player hits the normal persistent save OR initiates a persistent save via 'y' menu
- the save is made, then we call the read save function & show the player what they will get next time
- the player gets 3 options, "The information is good", "The information is not good; resave & reload it", "The information is not good repeatedly; create a log file"
- not sure if we snapshot every thing on the server or just memory/filesize profile/params size etc but generate something that the server admin can send to the devs
- save to a second spot in the file that all of the other code ignores.

Senario B:

We approach the situation from loading.

- Script to confirm after persistent save the data presented is what the player see's for their stats
- Leaving old save & load functions, etc in place
- If a player loads & is wiped they goto `y` menu and hit recovery from previous 
- Maybe call a script to detect if a player says yes to load previous & their stats are zeroed out - prompt them to reload from the backup (not the disconnect log)
- Game loads the 2nd'ary save that no other code touches so we know it's safe unless it's some arma3 engine bug overwriting etc

(if that works, I'll then submit a 2nd PR that abstracts your file loading to distinguish between normal & recovery mode to save code & vars ... perhaps even include a chunk to prepare for if you guys ever green light my SQL database stuff, but that's just bonus)

I'd rather not waste my time if you guys aren't open to submissions - so try & detail anything here that might be a road block.

**Comment by jaj22** [comment]
> do we think the issue is file writing, file max size, file loading, file var overwriting, memory over load?

None of the above. #1031 (distinguished by the loadout loading correctly and money being reset to 95 or 100) is caused by both the client and server using global setVariables during a player load, with understandably undefined behaviour. I have a proper fix for it but it's probably too much restructuring for 2.3.

#1198 (0 money, empty loadout) has a different cause, but it's still unlikely to be any of those, otherwise we'd get more general save corruption. The way player saves work is that money/score/rank/garage are stored as global variables on the player object, so most likely at the point when savePlayer accesses that data, it's incorrect or missing for some reason. Possibilities:

1. By the time the onPlayerDisconnect event handler is fired, the player object has changed for some reason.
2. The data on the player object was already desynchronized before the save happened.

If the cause is 1, the easy workaround is to detect whether the player object has the variables on it, and don't save if it doesn't. This would hopefully not be easily exploitable. I'll probably implement that, chuck in a log line and see if there's any pattern.

In general you don't want to give players a choice of which save to load, as that can be trivially exploited. As it's straightforward to Zeus or debug-console a player's personal save back into shape, it's not a huge problem anyway and not worth a ton of complexity to work around. It's just pissing me off that I don't know exactly why it's happening.

**Comment by Mirv** [comment]
Good stuff on the first fix!  I definitely feel the pain on state handling and clean up :(

> (0 money, empty loadout) 

Is the loadout you use here refer to arma3 load outs or to the stuff the player is carrying when they respawn?

I can't tell from saveplayer.sqf if you do want weapons on the player when they respawn or not?

I can see the code strips loadouts(inventory?) [fn_saveplayer.sqf](https://github.com/official-antistasi-community/A3-Antistasi/blob/981bf2ddb2be4d875b6f1bfd18838081a50b6a76/A3-Antistasi/functions/Save/fn_savePlayer.sqf).

I can also see in here that the [fn_loadplayer.sqf](url) loads the player?

I have regularly spawned with weapons if I had been alive when I logged out & persistent save.

I noticed that `stripGearFromLoadout` is only used in the single player...maybe we use this when loading players to prevent abuse of arsenal inventory?
